website/api/zkvm/input-output.md

```diff
@@ -0,0 +1,274 @@
+:::
+
+## Handling Inputs
+
+The [`stdin`][stdin] file descriptor is used to send input data from the host to
+the guest. The zkVM provides some convenience methods to read data from this
+file descriptor.
+
+- [`env::read`][read_method]: Reads a value from the `stdin` file descriptor.
+- [`env::read_slice`][read_slice_method]: Reads a slice from the `stdin` file
+  descriptor.
+
+For example, in the host program, we could:
+
+`/src/main.rs`
+
+```rust
+use risc0_zkvm::ExecutorEnv;
+
+fn main() {
+  let env = ExecutorEnv::builder()
+              // Write a single value to [stdin]
+              .write(42u64)
+              .unwrap()
+              // Write a slice to [stdin]
+              .write_slice(&[1u8, 2u8, 3u8])
+              .build();
+
+  // Do something with the ExecutorEnv...
+}
+```
+
+And in the guest program, we could:
+
+`/methods/guest/src/main.rs`
+
+```rust
+#![no_main]
+use risc0_zkvm::guest::env;
+risc0_zkvm::guest::entry!(main);
+
+fn main() {
+  // Read a single value from the standard input
+  let input: u64 = env::read()?;
+
+  // Read a slice from the standard input
+  let mut buffer = [0u8; 32];
+  env::read_slice(&mut buffer)?;
+
+  // Ensure we read the expected input
+  assert_eq!(input, 42);
+  assert_eq!(buffer, [1u8, 2u8, 3u8]);
+
+  // Do something with the input data...
+}
+```
+
+## Handling Private Outputs
+
+The [`stdout`][stdout] file descriptor is used to send private output from the
+guest to the host. The zkVM provides some convenience methods to write data to
+this file descriptor.
+
+- [`env::write`][write_method]: Writes a value to the `stdout` file descriptor.
+- [`env::write_slice`][write_slice_method]: Writes a slice to the `stdout` file
+  descriptor.
+
+Differently from the [`stdin`][stdin] file descriptor, in the [`stdout`][stdout]
+we need to provide a buffer to write the data to.
+
+The basic flow is that we need to instantiate a buffer in the host program, and
+then pass it to the guest program. The guest program will write the output to
+the [`stdout`][stdout] file descriptor, and the host program will read the
+output from the buffer.
+
+For example, in the host program, we could:
+
+`/src/main.rs`
+
+```rust
+use risc0_zkvm::{
+  ExecutorEnv,
+  Receipt,
+  default_prover,   // convenience method to create a default prover
+  serde::from_slice // necessary to deserialize the output
+};
+use app_methods::EXAMPLE_ELF;
+use app::ExampleStruct;
+
+fn main() {
+  // Step 1: Create a buffer to store the output
+  let mut output: Vec<u8> = Vec::new();
+  let env = ExecutorEnv::builder()
+              // Step 2: Instruct the guest to write to the desired buffer
+              .stdout(&mut output)
+              .build();
+  let prover = default_prover();
+  let receipt = prover.prove(env, EXAMPLE_ELF).unwrap();
+
+  // Step 3: Deserialize the output
+  let example_struct: ExampleStruct = from_slice(&output).unwrap();
+  let example_slice: [u8; 3] = from_slice(&output).unwrap();
```

Comment:
This will not do what's intended, because `from_slice` is a method for deserialization, and the data written with `write_slice` is not serialized.